KEYCLOAK-15524 Cleanup user related interfaces

mhajas opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mhajas committed:
KEYCLOAK-15524 Cleanup user related interfaces

hmlnarik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -237,35 +240,53 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
+     * <p/>
+     * Valid parameters are:
+     * <ul>
+     *     <li>"firstName" - first name (case insensitive string)</li>
```
Prefer constants, e.g. UserModel.FIRST_NAME to plain strings.
```suggestion
     *     <li>UserModel.FIRST_NAME - first name (case insensitive string)</li>
```

Same comment applies to methods below

hmlnarik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -323,14 +361,15 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
 
     /**
-     * Get users that belong to a specific group. Implementations do not have to search in UserFederatedStorageProvider
+     * Get users that belong to a specific group. Implementations do not have to search in {@code UserFederatedStorageProvider}
```
`Implementations do not have to search in {@code UserFederatedStorageProvider}` part should clarified / reworded

mhajas replied:
To be honest, I would probably completely remove this part, as it basically holds for all methods that they should not do anything with federated storage. Maybe it should be in the documentation, that when you are implementing UserProvider for your storage you should return only users, that are part of your storage.

hmlnarik replied:
I agree, class javadoc would be best place to put this information

hmlnarik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -63,39 +63,62 @@ default int getUsersCount(RealmModel realm, Set<String> groupIds) {
-     * @param search   search criteria
      * @param realm    the realm
+     * @param search   search criteria
      * @param groupIds set of groups to check for
```
What is checked in the set of groups?

hmlnarik commented on model/jpa/src/main/java/org/keycloak/models/jpa/PaginationUtils.java:
```diff
@@ -0,0 +1,18 @@
+
+public class PaginationUtils {
+    public static <T> TypedQuery<T> paginateQuery(TypedQuery<T> query, Integer first, Integer max) {
+        if (first != null && first > 0) {
```
Please note https://hibernate.atlassian.net/browse/HHH-14295

hmlnarik reviewed (changes requested):
Thanks @mhajas ! Looks almost good to me, mostly Javadoc needs fixing.

vramik commented on testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/UserPropertyFileStorage.java:
```diff
@@ -145,88 +148,59 @@ public int getUsersCount(RealmModel realm) {
+                .map(username -> createUser(realm, (String) username));
     }
-
+    
```
nit: forgotten white characters

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -432,21 +510,23 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
+     * @param realm a reference to the realm
+     * @return list of users with the given attribute name and value
+     *
+     * @deprecated Use {@link #searchForUserByUserAttributeStream(RealmModel, String, String) searchForUserByUserAttributeStream}
```
name

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -411,15 +487,17 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
+     * @param maxResults maximum number of results to return. Ignored if negative.
+     * @return a paginated list of users that has the given role assigned
+     *
      * @deprecated Use {@link #getRoleMembersStream(RealmModel, RoleModel, Integer, Integer) getRoleMembersStream} instead.
```
name

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -225,112 +263,148 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
+     * @param group a reference to the group
+     * @return a list of all users that are members of the given group
+     *
      * @deprecated Use {@link #getGroupMembersStream(RealmModel, GroupModel) getGroupMembersStream} instead.
```
name

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -352,16 +426,17 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
+     * @param maxResults maximum number of results to return. Ignored if negative.
+     * @return paginated list of members of the given group
+     *
      * @deprecated Use {@link #getGroupMembersStream(RealmModel, GroupModel, Integer, Integer) getGroupMembersStream} instead.
```
name

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -387,9 +462,10 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
+     * @param role a reference to the role
+     * @return a list of users that has the given role assigned
+     *
      * @deprecated Use {@link #getRoleMembersStream(RealmModel, RoleModel) getRoleMembersStream} instead.
```
name

vramik commented on server-spi/src/main/java/org/keycloak/models/GroupProvider.java:
```diff
@@ -140,7 +140,7 @@ default Long getGroupsCount(RealmModel realm, Stream<String> ids, String search)
      * @param maxResults Maximum number of results to return. Ignored if negative.
      * @return List of groups with the given role.
-     * @deprecated Use {@link #getGroupsByRoleStream(RealmModel, RoleModel, int, int) getGroupsByRoleStream} instead.
+     * @deprecated Use {@link #getGroupsByRoleStream(RealmModel, RoleModel, Integer, Integer)} instead.
```
name

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -154,7 +188,7 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
 
     /**
-     * @deprecated Use {@link #getUsersStream(RealmModel, int, int) getUsersStream} instead.
+     * @deprecated Use {@link #getUsersStream(RealmModel, Integer, Integer) getUsersStream} instead.
```
name

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -193,28 +229,30 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
+     * @param maxResults maximum number of results to return. Ignored if negative.
+     * @return paginated list of users from the realm that satisfies given search
+     *
+     * @deprecated Use {@link #searchForUserStream(RealmModel, String, Integer, Integer) searchForUserStream} instead.
```
name

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -225,112 +263,148 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
-     * @param realm
-     * @return
-     * @deprecated Use {@link #searchForUserStream(Map, RealmModel) searchForUserStream} instead.
+     * @deprecated Use {@link #searchForUserStream(RealmModel, Map) searchForUserStream} instead.
```
name

vramik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -225,112 +263,148 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
-     * @param maxResults
-     * @return
-     * @deprecated Use {@link #searchForUserStream(Map, RealmModel, Integer, Integer) searchForUserStream} instead.
+     * @deprecated Use {@link #searchForUserStream(RealmModel, Map, Integer, Integer) searchForUserStream} instead.
```
name

vramik reviewed (commented):
The PR looks really great, I have just few minor questions and it'll need a rebase as https://github.com/keycloak/keycloak/pull/7636 was merged so there will be conflicts.

mhajas commented on server-spi/src/main/java/org/keycloak/models/UserProvider.java:
```diff
@@ -88,104 +185,182 @@
+     *
+     * @param realm a reference to the realm
+     * @param user the user model
+     * @param socialProvider social provider id
```
This is very good point because it is actually `alias` that is used, not `internalId` or `providerId`.

mhajas commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -173,15 +173,16 @@ default int getUsersCount(RealmModel realm, boolean includeServiceAccount) {
 
     /**
-     * Search for users with username, email or first + last name that is like search string.
+     * Searches for users with username, email or first + last name that is like search string.
```
Should we do searching like described? As far as I know our implementations doesn't search for all words in `search` separately. The way I understand it, it just checks if any of strings `username`, `email` and `first name + last name` contains `search` string. See https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L824-L833. Am I missing something?

hmlnarik replied:
Yes. Building any sort of index on "first name + last" is hard and requires functional indices. It's easier to search word for word. JPA implementation is far from performant just because of this. For the fields alone, full-text search / index could be leveraged.

mhajas commented on federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java:
```diff
@@ -386,41 +379,34 @@ public int getUsersCount(RealmModel realm) {
     @Override
     public Stream<UserModel> getGroupMembersStream(RealmModel realm, GroupModel group, Integer firstResult, Integer maxResults) {
+        int first = firstResult == null ? 0 : firstResult;
+        int max = maxResults == null ? Integer.MAX_VALUE - 1 : maxResults;
```
Good question. I basically reused the default value that was there before. However, I am not sure it is correct, maybe we should create a Jira for this as [this](https://github.com/keycloak/keycloak/blob/master/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/MembershipType.java#L87) can cause integer overflow when `firstResult > 1`. @hmlnarik WDYT?

hmlnarik replied:
I'd say `Integer.MAX_VALUE >> 1` would do here for reasonable cases

mhajas commented on server-spi/src/main/java/org/keycloak/models/UserProvider.java:
```diff
@@ -38,39 +38,136 @@
+    UserModel getServiceAccount(ClientModel client);
+
+    /**
+     * @deprecated Use {@link #getUsersStream(RealmModel, boolean) getUsersStream} instead.
```
This is my mistake. I thought I just accidentally added the method at the end of `@link` block (I thought I pressed something what I should not when Intellij tries to hint the method name), however, it seems that the method name is there on purpose. Thanks for spotting this, I will add the method name everywhere, I think it is more readable when there is only method name, or WDYT?

mhajas commented on server-spi/src/main/java/org/keycloak/storage/user/UserLookupProvider.java:
```diff
@@ -27,16 +27,78 @@
  * @version $Revision: 1 $
  */
 public interface UserLookupProvider {
+
```
What do you mean? I think what is written there is still true.

vramik replied:
When I read it I realized I don't know how the interface `supports basic login capabilities` so I assumed it was outdated but it's probably just the fact that I don't know the application of the interface in those login capabilities.

mhajas replied:
I see. It works because when you log in, UserStorageManager will search all storage providers, that implement UserLookupProvider, for username/email. This means it should be enough to just implement UserLookupProvider and you should be able, to login with users stored in the storage. If you want some other functionality, for example, managing users via admin console, you need to implement also UserQueryProvider. After that, when UserStorageManager searches for all users, it will consider also users from your storage because you implemented UserQueryProvider. 

In code, this is handled by methods in [AbstractStorageManager](https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/storage/AbstractStorageManager.java), particularly in [getStorageProviderInstance(...)](https://github.com/keycloak/keycloak/blob/de09bdf32d6ede943bf2b0d60a9a452eef692129/services/src/main/java/org/keycloak/storage/AbstractStorageManager.java#L225) which will check, whether UserStorageProvider implements requested `capabilityInterface`.

vramik replied:
Thank for explanation.

mhajas commented on services/src/main/java/org/keycloak/storage/UserStorageManager.java:
```diff
@@ -161,10 +162,8 @@ protected void deleteInvalidUser(final RealmModel realm, final UserModel user) {
-        if (firstResult < 0) firstResult = 0;
-        if (maxResults < 0) maxResults = Integer.MAX_VALUE - 1;
+    protected Stream<UserModel> query(PaginatedQuery pagedQuery, RealmModel realm, Integer firstResult, Integer maxResults) {
+        if (maxResults != null && maxResults == 0) return Stream.empty();
```
This line will make sure that empty stream is returned when `maxResult == 0`. Even without this check the `paginatedStream` function will add `.limit(0)` because there is `>= 0`. I am not sure I understand what you mean.

vramik replied:
You're right, my mistake, please ignore.

mhajas commented on services/src/main/java/org/keycloak/storage/UserStorageManager.java:
```diff
@@ -362,54 +344,44 @@ public int getUsersCount(RealmModel realm, Set<String> groupIds) {
     }
 
     @Override // TODO: missing storageProviders count?
```
@hmlnarik should this be added in future or should I remove these comments?

hmlnarik replied:
Good catch @vramik. These methods need to fulfil the contract given in javadoc, and thus need to count in the federated storages as well

mhajas replied:
We should probably create a Jira for this, or should I add it in this PR?

hmlnarik replied:
Rather the latter (this PR) please.

vramik replied:
> [EDIT] @mhajas has convinced me that this is not a straightforward cleanup because of requirements it has on handling e.g. imported users and those from the original storage. Let's keep this in a separate task

So let's create a jira for that and leave it as it is.

mhajas replied:
Here is Jira: https://issues.redhat.com/browse/KEYCLOAK-16816

vramik commented on testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/UserPropertyFileStorage.java:
```diff
@@ -145,88 +148,59 @@ public int getUsersCount(RealmModel realm) {
-        }
-        return users;
+    public Stream<UserModel> getUsersStream(RealmModel realm, Integer firstResult, Integer maxResults) {
+        if (maxResults == 0) return Stream.empty();
```
Don't we need to ensure that `maxResult` is not `null`?

vramik reviewed (commented):
Thank you @mhajas! Great job. I've re-review your changes and I have one comment there. Other than that let's wait for @hmlnarik's view regarding those `//TODO`s and we should also probably run the pipelines.

hmlnarik commented on server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java:
```diff
@@ -29,9 +29,17 @@
- * if you want to view and manager users from the administration console.
+ * if you want to view and manage users from the administration console.
+ *
+ * Optional capability interface implemented by UserStorageProviders.
```
This note needs elaborating

The paragraphs need to be separated by `<p>`

mhajas commented:
Here are the pipeline results:
https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/2015/
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2490/

The results are not good, however, it seems it corresponds to previous runs from the stage branch, so none of the failure seems to be caused by this PR.

Thank you very much for your reviews @hmlnarik, @vramik! I think I addressed all comments and I think it should be ready to merge.

hmlnarik commented:
Thanks @mhajas ! Looks very good to me. The only two remaining points from my are the clarification of the {@code groupIds} parameter (https://github.com/keycloak/keycloak/pull/7673/files#r555339591) and squashing the commits. Great job!

mhajas commented:
@hmlnarik done.

hmlnarik reviewed (dismissed):
LGTM, thank you @mhajas!

martin-kanis reviewed (approved):
LGTM! Nice work, Michal.
